File: ExamThirdPartial/StrategySortingHidalgo.java/src/main/java/ec/edu/espe/controller/SortController.java
package ec.edu.espe.controller;

import ec.edu.espe.model.*;
import ec.edu.espe.view.SortView;
import org.bson.Document;
import com.mongodb.client.MongoClients;
import com.mongodb.client.MongoClient;
import com.mongodb.client.MongoDatabase;
import com.mongodb.client.MongoCollection;

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.util.Arrays;

public class SortController {
    private SortView view;
    private SortingContext context;
    
private static final String MONGO_URI = "mongodb+srv://Mikael:Jaime2006@cluster0.fpyoe9m.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0";
   
    private static final String DB_NAME = "strategyHidalgo"; 
    private static final String COLLECTION_NAME = "arrayMikael";

    public SortController(SortView view, SortingContext context) {
        this.view = view;
        this.context = context;
        
        this.view.addSortButtonListener(new SortListener());
    }

    class SortListener implements ActionListener {
        @Override
        public void actionPerformed(ActionEvent e) {
            try {
               
                String text = view.getInputText();
                if(text.isEmpty()) {
                    view.showMessage("Por favor ingresa números.");
                    return;
                }
                
           
                String[] parts = text.split(",");
                int[] numbers = new int[parts.length];
                for (int i = 0; i < parts.length; i++) {
                    numbers[i] = Integer.parseInt(parts[i].trim());
                }
                
                int n = numbers.length;
                
                if (n <= 1) {
                    view.showMessage("Debes ingresar al menos 2 números para ordenar.");
                    return;
                }
               
                if (n >= 2 && n <= 5) {
                    context.setSortStrategy(new BubbleSort());
                } 
               
                else if (n >= 6 && n <= 10) {
                    context.setSortStrategy(new InsertionSort());
                } 
               
                else { 
                    context.setSortStrategy(new QuickSort());
                }

                
                int[] sortedNumbers = context.sort(numbers);
                String algorithmUsed = context.getCurrentStrategyName();

            
                String resultMsg = "Algoritmo seleccionado automáticamente: " + algorithmUsed + "\n\n" +
                                   "Arreglo Original (Tamaño " + n + "):\n" + Arrays.toString(numbers) + "\n\n" +
                                   "Arreglo Ordenado:\n" + Arrays.toString(sortedNumbers);
                
                view.setResultText(resultMsg);
               
                view.showMessage(resultMsg);

               
                saveToMongoDB(numbers, sortedNumbers, algorithmUsed, n);

            } catch (NumberFormatException ex) {
                view.showMessage("Error: Asegúrate de ingresar solo números separados por comas válidas.");
            } catch (Exception ex) {
                ex.printStackTrace();
                view.showMessage("Ocurrió un error inesperado: " + ex.getMessage());
            }
        }
    }

    private void saveToMongoDB(int[] unsorted, int[] sorted, String algo, int size) {
        try (MongoClient mongoClient = MongoClients.create(MONGO_URI)) {
            MongoDatabase database = mongoClient.getDatabase(DB_NAME);
            MongoCollection<Document> collection = database.getCollection(COLLECTION_NAME);

            Document doc = new Document("unsorted", Arrays.toString(unsorted))
                    .append("size", size)
                    .append("sort algorithm", algo)
                    .append("sorted", Arrays.toString(sorted));

            collection.insertOne(doc);
            System.out.println("Registro guardado exitosamente en MongoDB Atlas.");
        } catch (Exception e) {
            view.showMessage("Error al conectar con MongoDB. Revisa tu conexión a internet o la cadena de conexión.\nDetalle: " + e.getMessage());
            e.printStackTrace();
        }
    }
}

File: ExamThirdPartial/ProyectoJsHidalgo/ec/edu/espe/controller/SortController.js
// ec/edu/espe/controller/SortController.js
const { MongoClient } = require('mongodb');
const { BubbleSort, InsertionSort, QuickSort, SortingContext } = require('../model/SortingModel');
const path = require('path');

// --- TU CONEXIÓN (Ya configurada) ---
const uri = "mongodb+srv://Mikael:Jaime2006@cluster0.fpyoe9m.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0";
const dbName = "strategyHidalgo";
const collectionName = "arrayMikael";

const controller = {};

// 1. Mostrar la página web (GET)
controller.showHome = (req, res) => {
    res.sendFile(path.join(__dirname, '../view/index.html'));
};

// 2. Procesar el ordenamiento (POST)
controller.processSort = async (req, res) => {
    try {
        const inputString = req.body.numbers;
        // Convertir texto "5,2,3" a array [5, 2, 3]
        const numbers = inputString.split(',').map(num => parseInt(num.trim())).filter(n => !isNaN(n));
        const n = numbers.length;

        if (n <= 1) {
            return res.send("Error: Ingresa al menos 2 números.");
        }

        // --- SELECCIÓN DE ESTRATEGIA (Lógica del Profesor) ---
        const context = new SortingContext();

        if (n >= 2 && n <= 5) {
            context.setStrategy(new BubbleSort());
        } else if (n >= 6 && n <= 10) {
            context.setStrategy(new InsertionSort());
        } else { // Mayor a 10
            context.setStrategy(new QuickSort());
        }

        // Ordenar
        const sortedNumbers = context.sort(numbers);
        const algoName = context.getStrategyName();

        // --- GUARDAR EN MONGODB ---
        const client = new MongoClient(uri);
        try {
            await client.connect();
            const db = client.db(dbName);
            const collection = db.collection(collectionName);

            const document = {
                unsorted: numbers.toString(),
                size: n,
                "sort algorithm": algoName,
                sorted: sortedNumbers.toString()
            };

            await collection.insertOne(document);
            console.log("Guardado en MongoDB Atlas exitosamente.");

        } catch (dbError) {
            console.error("Error Mongo:", dbError);
            return res.send("Error al conectar con la Base de Datos: " + dbError.message);
        } finally {
            await client.close();
        }

        // Enviar respuesta al navegador
        res.send(`
            <div style="font-family: sans-serif; text-align: center; padding: 50px;">
                <h1>Resultado Exitoso</h1>
                <p><strong>Algoritmo Usado:</strong> ${algoName}</p>
                <p><strong>Original:</strong> [${numbers}]</p>
                <p><strong>Ordenado:</strong> [${sortedNumbers}]</p>
                <p><em>Datos guardados en MongoDB (strategyHidalgo / arrayMikael)</em></p>
                <br>
                <a href="/"><button style="padding:10px; background:#007bff; color:white; border:none; cursor:pointer;">Volver</button></a>
            </div>
        `);

    } catch (e) {
        console.error(e);
        res.send("Ocurrió un error en el servidor: " + e.message);
    }
};

module.exports = controller;